lib: posix: Switch to use zephyr_interface_library_named cmake directive

pfalcon opened the pull request:
Similar to how other sub-libraries are defined in Zephyr tree, e.g.
"fs", "lgvl", etc. This is supposed to help with the need to
explicitly add posix include path to each and every application using
POSIX subsys.

Fixes: #15627

pfalcon committed:
lib: posix: Switch to use zephyr_interface_library_named cmake directive

Similar to how other sub-libraries are defined in Zephyr tree, e.g.
"fs", "lgvl", etc. This is supposed to help with the need to
explicitly add posix include path to each and every application using
POSIX subsys.

Fixes: #15627

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
tests: posix: No longer use target_include_directories

This is no longer needed after using
zephyr_interface_library_named cmake directive in the POSIX subsys
source itself.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

SebastianBoe commented:
We are also going to need a APP_LINK_WITH_PTHREAD Kconfig option as shown here:

https://github.com/zephyrproject-rtos/zephyr/commit/4c60c510c3350818b675d8ecc99dd1e46843321d

This is to allow the user to opt-out of this feature.

pfalcon commented:
These changes are for #15627, and currently don't work as expected (and thus, should not be merged). Discussion of the issue with these changes are at: https://github.com/zephyrproject-rtos/zephyr/issues/15627#issuecomment-489643141

pfalcon commented:
> We are also going to need a APP_LINK_WITH_PTHREAD Kconfig option as shown here:

So, current issue is that it simply doesn't work as expected. So, let's leave discussion of opt-out for later. (It would start with the fact that there's opt-in with CONFIG_POSIX_API in the first place.)

SebastianBoe commented:
Sorry, I meant the design-decision for why this option pattern exists is to allow user to opt-out of the feature.

The feature itself does require the option to be present for it to at all work.

This is documented in the usage documentation of the function here:

https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/extensions.cmake#L485

I see now that the wrong word is used in the documentation, it says "should", but should say "must".

pfalcon commented:
> The feature itself does require the option to be present for it to at all work.

Thanks, that helped, but I now see include order conflicts between libc vs POSIX headers (before they were in right order, now they're in wrong). Investigating...

pfalcon commented:
@PiotrZierhoffer, @tgorochowik: This is an example of trying to improve a little bit build process of POSIX apps.

pfalcon committed:
libc: minimal: Add headers as system includes

This is consistent with how newlib headers are treated, and will
have effect of ninlibc headers to be further down in the include
order. This is important, because some POSIX subsys headers
override those of libc. Without this change, we can't streamline
POSIX build config using zephyr_interface_library_named() cmake
directive, because includes will be in wrong order.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
CMakeLists.txt: Call toolchain_cc_nostdinc() at the end

To let any other zephyr_system_include_directories() directives
act first and get these paths in an order allowing Zephyr's stuff
override compiler headers.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

SebastianBoe commented on tests/posix/common/CMakeLists.txt:
```diff
@@ -4,6 +4,6 @@ cmake_minimum_required(VERSION 3.13.1)
 include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
 project(posix_common)
 
-target_include_directories(app PRIVATE $ENV{ZEPHYR_BASE}/include/posix)
```
Code should not be commented out.

pfalcon replied:
Fixed, and also removed this directive from the other posix test.

galak commented:
Sorry for taking a while to review, can you rebase and resolve the conflicts.

nashif commented on lib/posix/Kconfig:
```diff
@@ -96,4 +96,12 @@ config	POSIX_MAX_OPEN_FILES
 
+# The name of this option is mandated by zephyr_interface_library_named
+# cmake directive.
+config APP_LINK_WITH_PTHREAD
```
should be APP_LINK_WITH_POSIX

pfalcon replied:
> should be APP_LINK_WITH_POSIX

No, as explained by @SebastianBoe, this should match the existing config symbols, that's why it's APP_LINK_WITH_PTHREAD.

nashif replied:
which existing config symbols? If anything is still using PTHREAD, change it POSIX as well.

pfalcon replied:
> If anything is still using PTHREAD, change it POSIX as well.

But this is not subject of this change. The subject of this change is "lib: posix: Switch to use zephyr_interface_library_named cmake directive".

Exploration of more deep changes to POSIX subsys was done in e.g. https://github.com/zephyrproject-rtos/zephyr/pull/12984 . Feel free to merge it if you like it.

nashif replied:
> But this is not subject of this change. The subject of this change is "lib: posix: Switch to use zephyr_interface_library_named cmake directive".

it becomes a subject of this change when you introduce a new Kconfig using the old name and we will be stuck with it.

pfalcon replied:
> we will be stuck with it.

Why would we be stuck with? We'll change it (and hundreds of other mistakes and bugs in Zephyr) when it comes to it.

All in all, I'll defer to @galak, who assigned me to work on POSIX subsys improvements, regarding whether that should be part of this PR or not. In the meantime, working on other assignment from him, and can't work further on this, beyond keeping the build green.

pfalcon commented:
@galak 

> can you rebase and resolve the conflicts.

Done, build is green.

pfalcon commented:
Rebased on the latest master.

pfalcon committed:
lib: posix: Use "posix_subsys" as the CMake lib for the subsystem.

Historically, it used to be "PTHREAD", which is no longer true, as
POSIX subsys offers much more functionality than just Pthreads. Use
detailed name, like "posix_subsys", to avoid possible confusion with
ARCH_POSIX-related matters.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
> should be APP_LINK_WITH_POSIX

Ok, made this PR start with a patch to rename CMake library PTHREAD -> posix_subsys.

Rebased on master, CI passes. @nashif, as your feedback is addressed, and I'm not sure if you may be on vacation or away, dismissing your review requesting the rename.

pfalcon commented:
@carlescufi, @galak, @ioannisg: Please consider merging this long-waiting, long-approved PR.

ioannisg commented:
> 
> 
> @carlescufi, @galak, @ioannisg: Please consider merging this long-waiting, long-approved PR.

Thanks. No rush on this, anyways (it's a bug fix and can be done after feature freeze)
